File: src/main/go/aoc2019/IntComputer.go
package main

import (
	"fmt"
)

type OpCode int

const (
	HALT          OpCode = 99
	ADD           OpCode = 1
	MULT          OpCode = 2
	INPUT         OpCode = 3
	OUTPUT        OpCode = 4
	JUMP_IF_TRUE  OpCode = 5
	JUMP_IF_FALSE OpCode = 6
	LESS_THAN     OpCode = 7
	EQUALS        OpCode = 8
	RELBASE       OpCode = 9
)

// IntComputer executes AOC 2019 IntCode
type IntComputer struct {
	mem      [10000]int64
	iptr     int
	relBase  int64
	cntInstr int64

	data       []int64
	in         func() int64
	out        func(value int64)
	lastOutput int64

	inchan  chan int64
	outchan chan int64
	done    chan int64
}

func (c *IntComputer) run() {
	for !c.halted() {
		c.run1()
	}
	c.run1()
}

func (c *IntComputer) runToOutput() {
	for !c.halted() && OpCode(c.mem[c.iptr]) != OUTPUT {
		c.run1()
	}
	if !c.halted() {
		c.run1()
	}
}

func (c *IntComputer) halted() bool {
	return OpCode(c.mem[c.iptr]) == HALT
}

func (c *IntComputer) run1() {
	c.cntInstr++
	instr := c.mem[c.iptr]
	op := OpCode(instr % 100)
	p := c.mem[c.iptr : c.iptr+4]
	mode := []int{0, (int(instr) / 100) % 10, (int(instr) / 1000) % 10, (int(instr) / 10000) % 10}

	param := func(i int) int64 {
		switch mode[i] {
		case 1: // immediate
			return p[i]
		case 2: // relative
			return c.mem[p[i]+c.relBase]
		default: // position
			return c.mem[p[i]]
		}
	}

	tgt := func(i int) int64 {
		switch mode[i] {
		case 2: // relative
			return p[i] + c.relBase
		default: // position
			return p[i]
		}
	}

	switch op {
	case ADD:
		c.mem[tgt(3)] = param(1) + param(2)
		c.iptr += 4
	case MULT:
		c.mem[tgt(3)] = param(1) * param(2)
		c.iptr += 4
	case INPUT:
		c.mem[tgt(1)] = c.input()
		c.iptr += 2
	case OUTPUT:
		c.output(param(1))
		c.iptr += 2
	case JUMP_IF_TRUE:
		if param(1) != 0 {
			c.iptr = int(param(2))
		} else {
			c.iptr += 3
		}
	case JUMP_IF_FALSE:
		if param(1) == 0 {
			c.iptr = int(param(2))
		} else {
			c.iptr += 3
		}
	case LESS_THAN:
		if param(1) < param(2) {
			c.mem[tgt(3)] = 1
		} else {
			c.mem[tgt(3)] = 0
		}
		c.iptr += 4
	case EQUALS:
		if param(1) == param(2) {
			c.mem[tgt(3)] = 1
		} else {
			c.mem[tgt(3)] = 0
		}
		c.iptr += 4
	case RELBASE:
		c.relBase += param(1)
		c.iptr += 2
	case HALT:
		c.done <- c.lastOutput
		// fmt.Println("HALT")
	default:
		panic(fmt.Sprintf("unknown opcode in istruction %d", instr))
	}
}

func (c *IntComputer) output(val int64) {
	c.lastOutput = val
	if c.out != nil {
		c.out(val)
	} else {
		fmt.Println("output: ", val)
	}
}

func (c *IntComputer) input() int64 {
	v := int64(0)
	if c.in != nil {
		v = c.in()
	} else {
		v = c.data[0]
		c.data = c.data[1:]
	}
	// log.Println("input ", v)
	return v
}

// func (c *IntComputer) input1() int64 {
// 	v := <-c.in
// 	// log.Println("input ", v)
// 	return v
// }

func (c *IntComputer) provide(values ...int64) {
	c.data = append(c.data, values...)
}

func (c *IntComputer) init(program []int64) {
	for i := range c.mem {
		c.mem[i] = 0
	}
	copy(c.mem[0:], program)
	c.iptr = 0
	c.data = make([]int64, 0)
	c.done = make(chan int64, 1)
}

func (c *IntComputer) initAsync() {
	c.inchan = make(chan int64, 2)
	c.outchan = make(chan int64, 2)
	c.in = func() int64 { return <-c.inchan }
	c.out = func(v int64) { c.outchan <- v }
}

File: src/main/csharp/aoc2019/IntComputer.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace aoc2019
{
    enum OpCode { ADD = 1, MULT = 2, INPUT = 3, OUTPUT = 4, JUMP_IF_TRUE = 5, JUMP_IF_FALSE = 6, LESS_THAN = 7, EQUALS = 8, HALT = 99 }


    class IntComputer
    {
        public long[] Mem = new long[100000];
        int iptr = 0;
        bool Halted => Mem[iptr] == (int)OpCode.HALT;
        long instructionCount = 0;
        private readonly long[] program;

        public Queue<long> InputData {get;} = new Queue<long>();

        public IntComputer(long[] program)
        {
            this.program = program;
            Reset();
        }

        public void Run()
        {
            while (!Halted)
            {
                Run1();
            }
        }

        public void Run1()
        {
            instructionCount++;
            int instruction = (int)Mem[iptr];
            var op = (OpCode)(instruction % 100);

            (long p1, long p2, long p3) = (Mem[iptr + 1], Mem[iptr + 2], Mem[iptr + 3]);
            var mode = new int[] { 0, instruction / 100 % 10, instruction / 1000 % 10, instruction / 10000 % 10 };

            Func<int, long> param = i => mode[i] == 0 ? /*pos*/ Mem[Mem[iptr + i]] : /*imm*/ Mem[iptr + i];

            switch (op)
            {
                case OpCode.ADD:
                    Mem[p3] = param(1) + param(2);
                    iptr += 4;
                    break;

                case OpCode.MULT:
                    Mem[p3] = param(1) * param(2);
                    iptr += 4;
                    break;

                case OpCode.INPUT:
                    Mem[p1] = Input();
                    iptr += 2;
                    break;

                case OpCode.OUTPUT:
                    Output(param(1));
                    iptr += 2;
                    break;

                case OpCode.JUMP_IF_TRUE:
                    iptr = param(1) != 0 ? (int)param(2) : iptr + 3;
                    break;

                case OpCode.JUMP_IF_FALSE:
                    iptr = param(1) == 0 ? (int)param(2) : iptr + 3;
                    break;

                case OpCode.LESS_THAN:
                    Mem[p3] = (param(1) < param(2)) ? 1 : 0;
                    iptr += 4;
                    break;

                case OpCode.EQUALS:
                    Mem[p3] = (param(1) == param(2)) ? 1 : 0;
                    iptr += 4;
                    break;

                case OpCode.HALT:
                    break;

                default:
                    throw new NotImplementedException($"opcode {op} is unknown");
            }

        }

        public virtual void Output(long v)
        {
            Console.WriteLine("output: {0}", v);
        }

        public virtual long Input()
        {
            return InputData.Dequeue();
        }

        internal void Reset()
        {
            Array.Fill(Mem, 0);
            Array.Copy(program, Mem, program.Length);
            iptr = 0;
        }
    }
}